ADDR-132: Add 'Get Global Properties' proxy privilege

Seremba opened the pull request:
<!--- Add a pull request title above in this format -->
<!--- real example: 'ADDR-128 Add foreignKeyTableName to liquibase constraints' -->
<!--- 'ADDR-JiraIssueNumber JiraIssueTitle' -->
## Description of what I changed
<!--- Describe your changes in detail -->
<!--- It can simply be your commit message, which you must have -->


## Issue I worked on
<!--- This project only accepts pull requests related to open issues -->
<!--- Want a new feature or change? Discuss it in an issue first -->
<!--- Found a bug? Point us to the issue/or create one so we can reproduce it -->
<!--- Just add the issue number at the end: -->
see https://openmrs.atlassian.net/browse/ADDR-132

## Checklist: I completed these to help reviewers :)
<!--- Put an `x` in the box if you did the task -->
<!--- If you forgot a task please follow the instructions below -->
- [ ] My IDE is configured to follow the [**code style**](https://wiki.openmrs.org/display/docs/Java+Conventions) of this project.

  No? Unsure? -> [configure your IDE](https://wiki.openmrs.org/display/docs/How-To+Setup+And+Use+Your+IDE), format the code and add the changes with `git add . && git commit --amend`

- [ ] I have **added tests** to cover my changes. (If you refactored
  existing code that was well tested you do not have to add tests)

  No? -> write tests and add them to this commit `git add . && git commit --amend`

- [ ] I ran `mvn clean package` right before creating this pull request and
  added all formatting changes to my commit.

  No? -> execute above command

- [ ] All new and existing **tests passed**.

  No? -> figure out why and add the fix to your commit. It is your responsibility to make sure your code works.

- [ ] My pull request is **based on the latest changes** of the master branch.

  No? Unsure? -> execute command `git pull --rebase upstream master`

dkayiwa commented:
Can you look into the above failure on Java 17?

Seremba commented:
Hi @dkayiwa, I am done!

dkayiwa commented:
It has failed again.

Seremba commented:
Hello @dkayiwa , how about now?

Seremba commented:
Hi @dkayiwa, is there a way to test compatibility with a certain Java version before pushing code to github? I am currently using Java 8.

Seremba commented:
Hi @dkayiwa I am done making changes to this. Kindly review!

Seremba commented:
Hi @dkayiwa

Seremba commented:
Hi @dkayiwa, I set this to draft until the tests pass. However, currently, OpenMRS does not support Java 17, and this module affects other modules. Don't you think it is the cause for failures until maybe OpenMRS officially supports Java 17? What do you have to say?

dkayiwa commented:
> Don't you think it is the cause for failures until maybe OpenMRS officially supports Java 17? What do you have to say?

You are correct!

Seremba commented:
Thanks for the response. What do you suggest I should do?

dkayiwa commented:
> What do you suggest I should do?

Let us start by removing all the changes we had added for Java 17

Seremba commented:
> > What do you suggest I should do?
> 
> Let us start by removing all the changes we had added for Java 17

Hi @dkayiwa, I am done already

dkayiwa commented:
I still see more changes than i would expect.

Seremba commented:
I have followed [this](https://github.com/openmrs/openmrs-module-addresshierarchy/pull/41/commits) history of commits, tracking what I added and removed from the beginning and I don't seem to have left any change for Java 17. cc @dkayiwa

Seremba commented:
I am done @dkayiwa

dkayiwa commented:
Your pull request still have very many changes not related to the Get Global Properties privilege.

dkayiwa commented:
https://github.com/openmrs/openmrs-module-addresshierarchy/pull/41/files

Seremba commented:
Hi @dkayiwa, I am done.

dkayiwa commented on api/src/main/java/org/openmrs/module/addresshierarchy/service/AddressHierarchyServiceImpl.java:
```diff
@@ -703,21 +703,24 @@ else if (level.getParent() == null){
 			if ((this.fullAddressCacheInitialized == false || MapUtils.isEmpty(this.fullAddressCache))
-					&& this.getAddressHierarchyEntryCount() > 0) {
-
-				this.fullAddressCache = new HashMap<Locale, Map<String,List<String>> >();
```
Can we revert the formatting changes?

Seremba replied:
Okay. let me do that

Seremba commented:
Hi @dkayiwa I am done reverting the formatting changes

dkayiwa commented:
Do we need to remove the proxy privilege?

Seremba commented:
Do you mean with a try-finally block? I got an error with that and the problem did not get away.

Seremba commented:
Hi @dkayiwa , is there anything you require I do on this?

dkayiwa commented on api/src/main/java/org/openmrs/module/addresshierarchy/service/AddressHierarchyServiceImpl.java:
```diff
@@ -705,18 +705,26 @@ else if (level.getParent() == null){
-				getAddressesForLocale(locale);
-				this.fullAddressCacheInitialized = true;
+		try {
+			Context.addProxyPrivilege("Get Global Properties");
```
I would use a locally declared constant for the above string.

Seremba replied:
fine. Let me do that!

Seremba replied:
> I would use a locally declared constant for the above string.

Hi @dkayiwa, I am done with this!

dkayiwa commented:
Are you now able to run this module without errors in your modified openmrs-core?

Seremba commented:
> Are you now able to run this module without errors in your modified openmrs-core?

Yes. I can run the address hierarchy module without any errors.

dkayiwa commented:
Is it the only module that you have loaded? Or do you even have the rest of the reference application modules?

Seremba commented:
You mean loading and unloading the rest of the reference application modules?

dkayiwa commented:
I mean being able to run all the reference application modules with your changes in openmrs-core

Seremba commented:
> I mean being able to run all the reference application modules with your changes in openmrs-core

They also run well. I get no problem whatsoever with any reference application module.

dkayiwa commented:
How many reference application modules do you have loaded?

Seremba commented:
I have about 42 modules in the reference application
![Screenshot 2024-03-07 173220](https://github.com/openmrs/openmrs-module-addresshierarchy/assets/45463621/eaacf3e9-456c-41d8-9b28-f2743793aff7)
![Screenshot 2024-03-07 173242](https://github.com/openmrs/openmrs-module-addresshierarchy/assets/45463621/ad799b8d-7718-478b-a20c-37d60b754089)

dkayiwa commented:
So if those are successfully running and you have done all the required changes in openmrs-core as specified in that ticket, then you can ask for code review as advised here: https://openmrs.atlassian.net/wiki/spaces/docs/pages/25477199/Pull+Request+Tips

Seremba commented:
Thank you so much! Let me do that.
